### PR TITLE
feature(hardware): refactor hardware controller to watch for valid responses and error messages

### PR DESCRIPTION
**Pull request by ryanthecoder**
[RET-1250](https://opentrons.atlassian.net/browse/RET-1250)
[RET-1251](https://opentrons.atlassian.net/browse/RET-1251)

# Overview
This will let the hardware controller listen for and handle error messages in addition to regular acks
This also implements timeouts for messages that previously didn't respond. now there is a timeout if they don't respond with an Acknowledgment for those messages

# Changelog

# Review requests

# Risk assessment

### CODE REVIEW

File: api/src/opentrons/hardware_control/backends/ot3controller.py [hunk 1]
```diff
@@ -226,7 +226,11 @@ async def move(
         group = create_move_group(origin, moves, self._present_nodes, stop_condition)
         move_group, _ = group
         runner = MoveGroupRunner(move_groups=[move_group])
-        positions = await runner.run(can_messenger=self._messenger)
+        try:
+            positions = await runner.run(can_messenger=self._messenger)
+        except RuntimeError:
+            log.error("move group runner failed")
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/backends/ot3controller.py]: this isn't really a helpful error without more detail - let's either get rid of this `except/raise` and let higher levels handle it, or make this a `log.exception` that will automatically print the execption

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py [hunk 2]
```diff
@@ -34,6 +49,111 @@
...
+_Acks = List[_AckPacket]
+_AckIdFilter = [MessageId.acknowledgement, MessageId.error_message]
+
+_Basic_Nodes: List[NodeId] = [NodeId.gantry_x, NodeId.gantry_y, NodeId.head]
+_Gripper_SubNodes: List[NodeId] = [NodeId.gripper_g, NodeId.gripper_z]
+_Head_SubNodes: List[NodeId] = [NodeId.head_l, NodeId.head_r]
+
+
+class AcknowledgeListener:
+    """Helper clas for CanMessenger to listen for Acks back from commands."""
```
Comment (sfoster1) [inline hardware/opentrons_hardware/drivers/can_bus/can_messenger.py]:
```suggestion
    """Helper class for CanMessenger to listen for Acks back from commands."""
```

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py [hunk 3]
```diff
@@ -34,6 +49,111 @@
...
+    def __call__(
+        self, message: MessageDefinition, arbitration_id: ArbitrationId
+    ) -> None:
+        """Called by can messenger when a message arrives."""
+        if isinstance(message, Acknowledgement) or isinstance(message, ErrorMessage):
+            self.handle_ack(message, arbitration_id)
+
+    def handle_ack(self, message: _AckResponses, arbitration_id: ArbitrationId) -> None:
+        """Add the ack to the queue if it matches the message_index of the sent message."""
+        if message.payload.message_index == self._message.payload.message_index:
```
Comment (sfoster1) [inline hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:122]: can we massage things around a bit and maybe move this `if` to `__call__` or make it `if not (...): return` or something? it's tough to deal with long indented blocks

Comment (sfoster1) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:122]: In fact let's factor this into another function or set of functions honestly

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py [hunk 4]
```diff
@@ -146,13 +290,37 @@ async def _read_task(self) -> None:
...
+        elif error_payload.severity == ErrorSeverity.recoverable:
+            log.error(f"recived a firmware recoverable error {error_name}")
+        elif error_payload.severity == ErrorSeverity.unrecoverable:
+            log.critical(f"recived a firmware critical error {error_name}")
+
+        if error_payload.message_index == 0:
+            log.error(
+                f"error {str(err_msg)} recieved is asyncronous, raising exception"
+            )
+            raise RuntimeError("Async firmware error", str(err_msg))
```
Comment (sfoster1) [inline hardware/opentrons_hardware/drivers/can_bus/can_messenger.py]: let's make a new exception type for this that can hold the code and so on

File: hardware/opentrons_hardware/hardware_control/current_settings.py [hunk 5]
```diff
@@ -1,63 +1,81 @@
...
                 payload=payloads.MotorCurrentPayload(
                     hold_current=UInt32Field(int(currents[0] * (2**16))),
                     run_current=UInt32Field(int(currents[1] * (2**16))),
                 )
             ),
+            expected_nodes=[node],
         )
+        if error != ErrorCode.ok:
+            log.error(
+                f"recieved error {str(error)} trying to set currents for {str(node)}"
```
Comment (sfoster1) [inline hardware/opentrons_hardware/hardware_control/current_settings.py]:
```suggestion
                f"received error {str(error)} trying to set currents for {str(node)}"
```

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py [hunk 6]
```diff
@@ -146,13 +290,41 @@ async def _read_task(self) -> None:
...
 
+    def _handle_error(self, build: BinarySerializable) -> None:
+        err_msg = ErrorMessage(payload=build)  # type: ignore[arg-type]
+        error_payload: ErrorMessagePayload = err_msg.payload
+        error_name = ""
+        if error_payload.error_code.value in [err.value for err in ErrorCode]:
+            error_name = str(ErrorCode(error_payload.error_code.value).name)
+        else:
+            error_name = "UNKNOWN ERROR"
+        if error_payload.severity == ErrorSeverity.warning:
```
Comment (sfoster1) [inline hardware/opentrons_hardware/drivers/can_bus/can_messenger.py]: maybe make this block a `log_error` function?

File: hardware/opentrons_hardware/hardware_control/move_group_runner.py [hunk 7]
```diff
@@ -333,6 +344,19 @@ def _remove_move_group(
             log.info(f"Move group {group_id+self._start_at_index} has completed.")
             self._event.set()
 
+    def _handle_acknowledge(self, message: Acknowledgement) -> None:
+        log.debug("recieved ack")
+
+    def _handle_error(self, message: ErrorMessage) -> None:
+        self._recieved_error = message
+        if message.payload.severity.value == ErrorSeverity.warning:
```
Comment (sfoster1) [inline hardware/opentrons_hardware/hardware_control/move_group_runner.py]: could reuse that log function here

File: hardware/tests/opentrons_hardware/drivers/can_bus/test_can_messenger.py [hunk 8]
```diff
@@ -110,6 +113,102 @@ async def test_send(
...
+                    message_id=MessageId.acknowledgement,
+                    node_id=NodeId.host,
+                    function_code=0,
+                    originating_node_id=NodeId.gripper_g,
+                )
+            ),
+            data=message.payload.message_index.value.to_bytes(4, "big"),
+        )
+    )
+    print(message.payload.message_index.value.to_bytes(4, "big"))
```
Comment (sfoster1) [inline hardware/tests/opentrons_hardware/drivers/can_bus/test_can_messenger.py]: don't want to check in prints

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Couple refactor suggestions and typos and stuff but looks good overall

File: hardware/opentrons_hardware/drivers/can_bus/can_messenger.py (hunk 3 above)
Comment (ryanthecoder) [reply hardware/opentrons_hardware/drivers/can_bus/can_messenger.py:122]: didn't show up in these few lines of diff, but I broke it out into much nicer looking functions

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

**Review by Laura-Danielle — APPROVED** [review]
this also lgtm

**Pull request merged**